src/main/java/org/opensearch/neuralsearch/search/collector/PagingFieldCollector.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.search.collector;
+
+import java.io.IOException;
+import java.util.Objects;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.FieldDoc;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.Sort;
+import org.opensearch.common.Nullable;
+import org.opensearch.neuralsearch.search.HitsThresholdChecker;
+
+/**
+ * PagingFieldCollector collects the sorted results at the shard level for every individual query
+ * as per search_after criteria applied in the search request.
+ * It collects the list of TopFieldDocs.
+ */
+public final class PagingFieldCollector extends HybridTopFieldDocSortCollector {
+
+    private final Sort sort;
+    private final FieldDoc after;
+
+    public PagingFieldCollector(int numHits, HitsThresholdChecker hitsThresholdChecker, Sort sort, @Nullable FieldDoc after) {
+        super(numHits, hitsThresholdChecker, sort);
+        this.sort = sort;
+        this.after = after;
+    }
+
+    @Override
+    public LeafCollector getLeafCollector(LeafReaderContext context) {
+        docBase = context.docBase;
+        final int afterDoc = after.doc - docBase;
+        return new HybridTopDocSortLeafCollector(sort, after) {
+            @Override
+            public void collect(int doc) throws IOException {
+                if (Objects.isNull(compoundQueryScorer)) {
+                    throw new IllegalArgumentException("scorers are null for all sub-queries in hybrid query");
+                }
+                float[] subScoresByQuery = compoundQueryScorer.hybridScores();
+                initializePriorityQueuesWithComparators(context, subScoresByQuery.length);
+                incrementTotalHitCount();
+                for (int i = 0; i < subScoresByQuery.length; i++) {
+                    float score = subScoresByQuery[i];
+                    // if score is 0.0 there is no hits for that sub-query
+                    if (score == 0) {
+                        continue;
+                    }
+
+                    // if queueFull[i] is true then it indicates
+                    // that we have found the results equal to the size sent in the search request.
+                    if (queueFull[i]) {
+                        // If threshold is reached then return. Default value of threshold is 10000.
+                        if (thresholdCheck(doc, i)) {
+                            return;
+                        }
+                    }
```

Comment:
why we are doing this check here before knowing result is on previous page or not? plus this is not same as SimpleFieldCollector why is that?

Reply:
Lets understand with an example
Consider there is a search_after condition to fetch results after value 40. Now from the total results 30 results fall after value 30. 
ie. 41......70

Now in the search query I have give the results size = 10

The final result should be from 41....50. 

Now answering your question

```
 if (queueFull[i]) {
                        // If threshold is reached then return. Default value of threshold is 10000.
                        if (thresholdCheck(doc, i)) {
                            return;
                        }
                    }
```

This code segment checks for that results size i.e I got 10 results already. 

Then we check whether all results has to be collected or not. i.e If I haven't found 10 results then continue the collection of docId's

Now if  result size is let say 100 and as I mentioned earlier the total results are 30. So it will continue iterating and the above condition of queue[i] will not fulfilled therefore we will keep on iterating.

Furthermore, answering your second concern the SimpleFieldCollector is just responsible for sorting the results. Therefore if the the queue is full it means I have found the desired number of results needed from size param sent in the search request. 


I hope that clears everything.